### PR TITLE
Add Event Image Support with validation

### LINKED ISSUE #1645: Add Support for Event Images
**Is your feature request related to a problem? Please describe.**

We need the ability to apply images to events. This could be useful in making the event more interesting because of an added flyer.

**Describe the solution you'd like**

1. We need an `image`  field added to the Event collection
1. You will be required to write tests for the expected CRUD activities related to the image.

**Describe alternatives you've considered**
- N/A

**Approach to be followed (optional)**
- See above

**Additional context**

The current screen shot of the `develop` branch's schema

image

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by duplixx**
**What kind of change does this PR introduce?**

feature

**Issue Number:**

Fixes #1645 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

image
image

**If relevant, did you update the documentation?**

**Summary**

**Does this PR introduce a breaking change?**

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

### CODE REVIEW

**Comment by duplixx** [comment]
@palisadoes fixed some type failing issues

**Comment by duplixx** [comment]
@JamaicanFriedChicken @sumitra19jha @palisadoes can you approve workflows to run 
thanks !!

File: schema.graphql [hunk 1]
```diff
@@ -263,6 +267,20 @@ enum EventOrderByInput {
   title_DESC
 }
 
+type EventProject {
```
Comment (xoldyckk) [inline schema.graphql:270]: @palisadoes wasn't this removed?

Comment (palisadoes) [reply schema.graphql:270]: Yes it should have been

Comment (palisadoes) [reply schema.graphql:270]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 2]
```diff
@@ -831,6 +856,37 @@ type Subscription {
   onPluginUpdate: Plugin
 }
 
+type Task {
+  _id: ID!
+  completed: Boolean
+  createdAt: DateTime!
+  creator: User!
```
Comment (palisadoes) [inline schema.graphql:863]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 3]
```diff
@@ -831,6 +856,37 @@ type Subscription {
...
+  deadline: DateTime
+  description: String
+  event: Event!
+  title: String!
+  volunteers: [User]
+}
+
+input TaskInput {
+  deadline: DateTime!
+  description: String!
```
Comment (palisadoes) [inline schema.graphql:873]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 4]
```diff
@@ -831,6 +856,37 @@ type Subscription {
...
+  title: String!
+}
+
+enum TaskOrderByInput {
+  createdAt_ASC
+  createdAt_DESC
+  deadline_ASC
```
Comment (palisadoes) [inline schema.graphql:880]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 5]
```diff
@@ -877,14 +934,26 @@ input UpdateEventInput {
   title: String
 }
 
+input UpdateEventProjectInput {
+  description: String
+  title: String
```
Comment (palisadoes) [inline schema.graphql:939]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 6]
```diff
@@ -877,14 +934,26 @@ input UpdateEventInput {
...
+  isPublic: Boolean
   location: String
   name: String
-  userRegistrationRequired: Boolean
   visibleInSearch: Boolean
 }
 
+input UpdateTaskInput {
+  completed: Boolean
+  deadline: DateTime
```
Comment (palisadoes) [inline schema.graphql:952]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: src/typeDefs/types.ts [hunk 7]
```diff
@@ -120,10 +121,19 @@ export const types = gql`
...
+    projects: [EventProject]
     feedback: [Feedback!]!
     averageFeedbackScore: Float
   }
 
+  type EventProject {
+    _id: ID!
+    title: String!
+    description: String!
+    event: Event!
```
Comment (palisadoes) [inline src/typeDefs/types.ts:133]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: src/typeDefs/types.ts [hunk 8]
```diff
@@ -311,6 +331,18 @@ export const types = gql`
     aggregate: AggregatePost!
   }
 
+  type Task {
+    _id: ID!
+    title: String!
+    description: String
+    event: Event!
+    creator: User!
```
Comment (palisadoes) [inline src/typeDefs/types.ts:339]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: src/typeDefs/types.ts [hunk 9]
```diff
@@ -271,6 +281,16 @@ export const types = gql`
     uninstalledOrgs: [ID!]!
   }
 
+  # type Plugin {
+  #   orgId: Organization!
+  #   pluginName: String!
+  #   pluginKey: String
+  #   pluginStatus: Status!
```
Comment (palisadoes) [inline src/typeDefs/types.ts:288]: Why was this removed?

File: schema.graphql [hunk 10]
```diff
@@ -525,10 +548,12 @@ type Mutation {
   unlikePost(id: ID!): Post
   unregisterForEventByUser(id: ID!): Event!
   updateEvent(data: UpdateEventInput, id: ID!): Event!
+  updateEventProject(data: UpdateEventProjectInput!, id: ID!): EventProject!
```
Comment (palisadoes) [inline schema.graphql:551]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 11]
```diff
@@ -525,10 +548,12 @@ type Mutation {
...
   updateLanguage(languageCode: String!): User!
   updateOrganization(data: UpdateOrganizationInput, file: String, id: ID!): Organization!
   updatePluginStatus(id: ID!, orgId: ID!): Plugin!
   updatePost(data: PostUpdateInput, id: ID!): Post!
+  updateTask(data: UpdateTaskInput!, id: ID!): Task
```
Comment (palisadoes) [inline schema.graphql:556]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 12]
```diff
@@ -517,6 +539,7 @@ type Mutation {
   sendMembershipRequest(organizationId: ID!): MembershipRequest!
   sendMessageToDirectChat(chatId: ID!, messageContent: String!): DirectChatMessage!
   sendMessageToGroupChat(chatId: ID!, messageContent: String!): GroupChatMessage!
+  setTaskVolunteers(id: ID!, volunteers: [ID]!): Task
```
Comment (palisadoes) [inline schema.graphql:542]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 13]
```diff
@@ -501,13 +521,15 @@ type Mutation {
...
   removeEventAttendee(data: EventAttendeeInput!): User!
+  removeEventProject(id: ID!): EventProject!
   removeGroupChat(chatId: ID!): GroupChat!
   removeMember(data: UserAndOrganizationInput!): Organization!
   removeOrganization(id: ID!): User!
   removeOrganizationCustomField(customFieldId: ID!, organizationId: ID!): OrganizationCustomField!
   removeOrganizationImage(organizationId: String!): Organization!
   removePost(id: ID!): Post
   removeSampleOrganization: Boolean!
+  removeTask(id: ID!): Task
```
Comment (palisadoes) [inline schema.graphql:532]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 14]
```diff
@@ -501,13 +521,15 @@ type Mutation {
   removeDirectChat(chatId: ID!, organizationId: ID!): DirectChat!
   removeEvent(id: ID!): Event!
   removeEventAttendee(data: EventAttendeeInput!): User!
+  removeEventProject(id: ID!): EventProject!
```
Comment (palisadoes) [inline schema.graphql:524]: etc

File: schema.graphql [hunk 15]
```diff
@@ -472,13 +490,15 @@ type Mutation {
...
   createEvent(data: EventInput): Event!
+  createEventProject(data: EventProjectInput!): EventProject!
   createGroupChat(data: createGroupChatInput!): GroupChat!
   createMember(input: UserAndOrganizationInput!): Organization!
   createMessageChat(data: MessageChatInput!): MessageChat!
   createOrganization(data: OrganizationInput, file: String): Organization!
   createPlugin(pluginCreatedBy: String!, pluginDesc: String!, pluginName: String!, uninstalledOrgs: [ID!]): Plugin!
   createPost(data: PostInput!, file: String): Post
   createSampleOrganization: Boolean!
+  createTask(data: TaskInput!, eventProjectId: ID!): Task!
```
Comment (palisadoes) [inline schema.graphql:501]: etc

File: schema.graphql [hunk 16]
```diff
@@ -472,13 +490,15 @@ type Mutation {
   createDirectChat(data: createChatInput!): DirectChat!
   createDonation(amount: Float!, nameOfOrg: String!, nameOfUser: String!, orgId: ID!, payPalId: ID!, userId: ID!): Donation!
   createEvent(data: EventInput): Event!
+  createEventProject(data: EventProjectInput!): EventProject!
```
Comment (palisadoes) [inline schema.graphql:493]: etc

File: schema.graphql [hunk 17]
```diff
@@ -263,6 +267,20 @@ enum EventOrderByInput {
...
+  _id: ID!
+  description: String!
+  event: Event!
+  tasks: [Task]
+  title: String!
+}
+
+input EventProjectInput {
+  description: String!
+  eventId: ID!
```
Comment (palisadoes) [inline schema.graphql:280]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

File: schema.graphql [hunk 18]
```diff
@@ -209,6 +210,7 @@ type Event {
   location: String
   longitude: Longitude
   organization: Organization
+  projects: [EventProject]
```
Comment (palisadoes) [inline schema.graphql:213]: @duplixx 

1. We removed Projects and Tasks from the API a few PRs ago.  Your starting code is out of date
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1627
1. This is going to be replaced with Action Items.
    - https://github.com/PalisadoesFoundation/talawa-api/pull/1648
1. Remove this Type and all references to it

**Review by palisadoes — CHANGES_REQUESTED** [review]
@duplixx It will be easier for you to start with the most recent upstream and add your changes. You are re-introducing too many deprecated features.

DO NOT CLOSE THIS PR. Make the appropriate updates.

**Comment by duplixx** [comment]
@palisadoes can i get review

**Comment by palisadoes** [comment]
Please review, this is ready.

- @JamaicanFriedChicken
- @sumitra19jha
- @xoldyckk

**Comment by duplixx** [comment]
hi @xoldyckk @JamaicanFriedChicken @sumitra19jha can you please review it so that I can start working on front end implementation

**Comment by sumitra19jha** [comment]
@palisadoes I need some requirement clarification before reviewing.

1. Is the scope of this issue is also to add image (into some blob storage like s3)?
2. Is there any restrictions on the size of image?
3. Is there any restrictions on the type of image (png, jpeg, jpg)
4. Are we storing blob storage path location or just the image name?


Based on this I will be able to review the PR

**Comment by palisadoes** [comment]
@sumitra19jha 


1. Is the scope of this issue is also to add image (into some blob storage like s3)?
    1.  We have a draft GSoC idea to solve object storage. The current default methodology should be used for now.
1. Is there any restrictions on the size of image?
    1. There will need to be. It should be out of scope for the purposes of this PR. I'll open another issue to address this for all image uploads
1. Is there any restrictions on the type of image (png, jpeg, jpg)
    1. We should limit it to the most common image format standards
1. Are we storing blob storage path location or just the image name?
    1.  The current default methodology should be used for now.

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by duplixx** [comment]
Hey @palisadoes, all test cases are now passing, but I am facing a little issue. When I am defining the error to be 'any', it gives me an error. However, in another file, the same is done without any errors. Here is the

 image
 
 
image

**Comment by duplixx** [comment]
issue resolved pushed the changes

**Comment by palisadoes** [comment]
Please address the test code coverage failure. Your patch is significantly below.

**Comment by duplixx** [comment]
@Cioppolo14 @sumitra19jha can you approve the workflow

**Comment by duplixx** [comment]
@Cioppolo14 Sorry to ask you again for the workflow approval, but this issue needs to be merged quickly. 😅

**Comment by duplixx** [comment]
these errors are not coming from my changes..what should i do to fix it
image

**Comment by duplixx** [comment]
If i fix using `npm run format:fix` following errors are coming

image

**Comment by palisadoes** [comment]
Please do a minor commit. I made a change that hopefully should fix the linting GitHub action.

**Comment by duplixx** [comment]
@sumitra19jha  can you approve workflow

**Comment by duplixx** [comment]
Hello @palisadoes, despite running Prettier and fetching the upstream changes, I'm still encountering errors in the code. Could you please guide how to resolve them?

Screenshot 2024-02-11 at 8 56 02 PM

**Comment by palisadoes** [comment]
1. We had to revert the prettier package upgrade last week due to unexpected errors.
2. Merge with the latest upstream

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by palisadoes** [comment]
@duplixx Are you still working on this?

**Comment by duplixx** [comment]
yes i am working on it..will do the required changes and fix the conflicts

File: src/typeDefs/types.ts [hunk 19]
```diff
@@ -164,7 +164,8 @@ export const types = gql`
     title: String!
     description: String!
     startDate: Date!
-    endDate: Date
+    endDate: Date!
```
[the remark this answers is not in the document]
Comment (duplixx) [reply src/typeDefs/types.ts]: thanks! have updated

[the remark this answers is not in the document]
Comment (meetulr) [reply src/typeDefs/types.ts]: thank you!

**Comment by meetulr** [comment]
@duplixx brother, I'm just a contributor, I just noticed one thing and commented on it, didn't know it would start a review😅.

**Comment by duplixx** [comment]
@Cioppolo14  can you approve the workflow

**Comment by Cioppolo14** [comment]
@duplixx Are you working on fixing the Testing Application failed check?

**Comment by duplixx** [comment]
yes working on it

**Comment by duplixx** [comment]
image
@JamaicanFriedChicken @xoldyckk  @sumitra19jha all test passed yesterday but when i updated my branch these new errors came as output,do we alos have to fix these also?

**Comment by duplixx** [comment]
@sumitra19jha @JamaicanFriedChicken can this PR get reviewed

**Comment by duplixx** [comment]
@palisadoes need review here

**Comment by palisadoes** [comment]
1. Weekday reviews are going to be slow due to work commitments. 
2. We'll send reminders tomorrow

**Review by palisadoes — APPROVED** [review]
